File: src/lib/relays/subscription-id.ts
import { sha256 } from '@noble/hashes/sha2.js';
import { bytesToHex } from '@noble/hashes/utils.js';
import type { FeedCursorPoint } from '../events/types';

export const maxRelaySubscriptionIdLength = 48;

export function relaySubscriptionHash(value: unknown, length = 12): string {
  const text = typeof value === 'string' ? value : JSON.stringify(value);
  return bytesToHex(sha256(new TextEncoder().encode(text))).slice(0, length);
}

export function compactRelaySubscriptionId(
  prefix: string,
  topic: string,
  discriminator?: unknown,
): string {
  const parts = discriminator
    ? [prefix, topic, relaySubscriptionHash(discriminator)]
    : [prefix, topic];
  const id = parts.join(':');
  if (id.length <= maxRelaySubscriptionIdLength) return id;
  return `${prefix.slice(0, 8)}:${topic.slice(0, 10)}:${relaySubscriptionHash(id, 16)}`;
}

export function childRelaySubscriptionId(
  parent: string,
  topic: string,
  discriminator?: unknown,
): string {
  return compactRelaySubscriptionId(parent, topic, discriminator);
}

export function liveRelaySubscriptionId(prefix: string, topic: string): string {
  return compactRelaySubscriptionId(prefix, topic);
}

export function initialRelaySubscriptionId(
  prefix: string,
  discriminator?: unknown,
): string {
  return compactRelaySubscriptionId(prefix, 'initial', discriminator);
}

export function olderRelaySubscriptionId(
  prefix: string,
  cursor: FeedCursorPoint | number,
): string {
  return compactRelaySubscriptionId(prefix, 'older', cursor);
}

export function newerRelaySubscriptionId(
  prefix: string,
  cursor: FeedCursorPoint | number,
): string {
  return compactRelaySubscriptionId(prefix, 'newer', cursor);
}

export function relaySubscriptionIdValid(id: string): boolean {
  return id.length > 0 && id.length <= maxRelaySubscriptionIdLength;
}

File: crates/lkjstr-relays/src/subscription_id.rs
#![doc = "Relay subscription id helpers."]

use sha2::{Digest, Sha256};

pub const fn max_relay_subscription_id_length() -> usize {
    48
}

#[must_use]
pub fn relay_subscription_hash(value: &str, length: usize) -> String {
    let digest = Sha256::digest(value.as_bytes());
    let hex = bytes_to_hex(&digest);
    let end = length.min(hex.len());
    hex[..end].to_owned()
}

#[must_use]
pub fn compact_relay_subscription_id(
    prefix: &str,
    topic: &str,
    discriminator: Option<&str>,
) -> String {
    let id = match discriminator {
        Some(value) if !value.is_empty() => {
            format!("{prefix}:{topic}:{}", relay_subscription_hash(value, 12))
        }
        _ => format!("{prefix}:{topic}"),
    };
    if id.len() <= max_relay_subscription_id_length() {
        return id;
    }
    format!(
        "{}:{}:{}",
        prefix.chars().take(8).collect::<String>(),
        topic.chars().take(10).collect::<String>(),
        relay_subscription_hash(&id, 16)
    )
}

#[must_use]
pub fn child_relay_subscription_id(
    parent: &str,
    topic: &str,
    discriminator: Option<&str>,
) -> String {
    compact_relay_subscription_id(parent, topic, discriminator)
}

#[must_use]
pub fn live_relay_subscription_id(prefix: &str, topic: &str) -> String {
    compact_relay_subscription_id(prefix, topic, None)
}

#[must_use]
pub fn initial_relay_subscription_id(prefix: &str, discriminator: Option<&str>) -> String {
    compact_relay_subscription_id(prefix, "initial", discriminator)
}

#[must_use]
pub fn older_relay_subscription_id(prefix: &str, cursor: &str) -> String {
    compact_relay_subscription_id(prefix, "older", Some(cursor))
}

#[must_use]
pub fn newer_relay_subscription_id(prefix: &str, cursor: &str) -> String {
    compact_relay_subscription_id(prefix, "newer", Some(cursor))
}

#[must_use]
pub fn relay_subscription_id_valid(id: &str) -> bool {
    !id.is_empty() && id.len() <= max_relay_subscription_id_length()
}

fn bytes_to_hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        out.push(hex_char(byte >> 4));
        out.push(hex_char(byte & 0x0f));
    }
    out
}

const fn hex_char(value: u8) -> char {
    match value {
        0..=9 => (b'0' + value) as char,
        _ => (b'a' + value - 10) as char,
    }
}
